File: export.py
import csv
import json
import os
import requests
import sys


def fetch_reader_document_list_api(token, updated_after=None, location=None):
    full_data = []
    next_page_cursor = None
    while True:
        params = {"withHtmlContent": False, "withRawSourceUrl": False}
        if next_page_cursor:
            params["pageCursor"] = next_page_cursor
        if updated_after:
            params["updatedAfter"] = updated_after
        if location:
            params["location"] = location
        response = requests.get(
            url="https://readwise.io/api/v3/list/",
            params=params,
            headers={"Authorization": f"Token {token}"},
            # verify=False,
        )
        full_data.extend(response.json()["results"])
        next_page_cursor = response.json().get("nextPageCursor")
        if not next_page_cursor:
            break
    return full_data


READWISE_READER_API_TOKEN = os.environ.get("READWISE_READER_API_TOKEN")
if not READWISE_READER_API_TOKEN:
    sys.stderr.write("READWISE_READER_API_TOKEN environment variable required")
    sys.exit(1)

all_data = fetch_reader_document_list_api(READWISE_READER_API_TOKEN)
print(json.dumps(all_data))

File: export.ts
declare module "bun" {
  interface Env {
    READWISE_READER_API_TOKEN: string;
  }
}

type Document = {
    id: string,
    url: string,
    title: string,
    author: string,
    source: string,
    category: string,
    location: string,
    tags: {},
    site_name: string,
    word_count: number | null,
    reading_time: number | null,
    created_at: string,
    updated_at: string,
    published_date: number,
    summary: string,
    image_url: string,
    content: string | null,
    source_url: string,
    notes: string,
    parent_id: string | null,
    reading_progress: number,
    first_opened_at: string | null,
    last_opened_at: string | null,
    saved_at: string | null,
    last_moved_at: null
}

const fetchDocumentListApi = async (updatedAfter: string | null=null, location: string | null=null) => {
    let fullData: Array<Document> = [];
    let nextPageCursor: string | null = null;

    while (true) {
      const queryParams = new URLSearchParams();
      if (nextPageCursor) {
        queryParams.append('pageCursor', nextPageCursor);
      }
      if (updatedAfter) {
        queryParams.append('updatedAfter', updatedAfter);
      }
      if (location) {
        queryParams.append('location', location);
      }
      const response = await fetch('https://readwise.io/api/v3/list/?' + queryParams.toString(), {
        method: 'GET',
        headers: {
          Authorization: `Token ${token}`,
        },
      });
      const responseJson = await response.json();
      fullData.push(...responseJson['results']);
      nextPageCursor = responseJson['nextPageCursor'];
      if (!nextPageCursor) {
        break;
      }
    }
    return fullData;
};

const token = process.env.READWISE_READER_API_TOKEN;
if(!token) {
    console.log("READWISE_READER_API_TOKEN environment variable required");
    process.exit(1);
}

// Get all of a user's documents from all time
const allData = await fetchDocumentListApi();
console.log(JSON.stringify(allData));